### PR TITLE
Generate rules with Python

### PR DESCRIPTION
**Pull request by robin-nitrokey**
This patch adds a Python script to generate the udev rules.  The required data is defined in a TOML file.  This makes the rules easier to maintain and update.

This causes some small changes in the formatting of the rules.  Also, the rule for the Nitrokey Storage bootloader is updated to use ATTRS instead of ATTR for consistency with similar rules.

### CODE REVIEW

**Review by daringer — APPROVED** [review]

File: generate.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+    if ccid_devices:
+        sections.append(generate_ccid(ccid_devices))
+
+    output = textwrap.dedent(header)
+    output += "\n\n".join(sections)
+    # TODO: can we remove this?
+    output += textwrap.dedent("""
+
+        # Nitrokey Storage dev Entry
+        KERNEL=="sd?1", ATTRS{idVendor}=="20a0", ATTRS{idProduct}=="4109", SYMLINK+="nitrospace"
```
Comment (robin-nitrokey) [inline generate.py:102]: I forgot to mention – this rule looks somewhat misplaced.  I understand it’s for development?  Should we just remove it from the official rules?

Comment (jans23) [reply generate.py:102]: @szszszsz Any idea, opinion?

**Review by sosthene-nitrokey — APPROVED** [review]

**Comment by robin-nitrokey** [comment]
Merged with the unclear Nitrokey Storage rule to keep the status quo.  We can still remove it in another PR.

**Comment by szszszsz** [comment]
Hi!

It's a feature actually - this makes a symlink to a first partition of the
connected Nitrokey Storage, regardless of what block device it was assigned
to (so nitrospace -> sda1, if Nitrokey Storage was assigned to sda, sdb1 if
sdb etc.). This way one could always run `mount /dev/nitrospace
/mnt/nitrokey`.

I've never heard anyone is using this one though, so I assume it's safe to
remove (perhaps I was using this internally). Now with the volume/partition
label based mounting this could be a little outdated as well.

Note, that this rule works just for a single device case. It would need to
be corrected for the multiple Nitrokey Storage devices connected.


pon., 16 gru 2024 o 08:27 jans23 ***@***.***> napisał(a):

**Comment by robin-nitrokey** [comment]
@szszszsz Thanks for the explanation!  Let’s remove it from the general rules then.  Users can still add a custom rule based on their setup and device.

**Pull request merged**